### PR TITLE
Deprecate it_comest [RDY]

### PR DESCRIPTION
**Pull request by mugling**
Depends upon  #15901 and #15908 (which are ready to merge) and will be rebased after those are merged.

New commits start with d26e3e2

### CODE REVIEW

**Comment by mugling** [comment]
a8e1a16 drops all `virtual` functions except the destructor which is required for the artifact types (which a separate PR will refactor).

**Comment by Coolthulhu** [comment]
Needs a rebase.

**Comment by mugling** [comment]
> Needs a rebase.

Yes, I'm going to wait for #15901 and then rebase after

**Comment by mugling** [comment]
Rebased against `master`

File: src/item.cpp [hunk 1]
```diff
@@ -733,16 +724,14 @@ std::string item::info( bool showtext, std::vector<iteminfo> &info ) const
...
+        if( g->u.nutrition_for( food_item->type ) != 0 || food_item->type->comestible->quench != 0 ) {
+            info.push_back( iteminfo( "FOOD", _( "<bold>Nutrition</bold>: " ), "", g->u.nutrition_for( food_item->type ),
                                       true, "", false, true ) );
-            info.push_back( iteminfo( "FOOD", space + _( "Quench: " ), "", food->quench ) );
+            info.push_back( iteminfo( "FOOD", space + _( "Quench: " ), "", food_item->type->comestible->quench ) );
         }
 
-        if( food->fun != 0 ) {
-            info.push_back( iteminfo( "FOOD", _( "Enjoyability: " ), "", food->fun ) );
+        if( food_item->type->comestible->fun ) {
```
Comment (Coolthulhu) [inline src/item.cpp]: Comparing to 0 is better than casting to bool

File: src/item_factory.cpp [hunk 2]
```diff
@@ -1002,29 +1004,44 @@ void Item_factory::load_book( JsonObject &jo )
...
+    jo.read( "tool", slot.tool );
+    jo.read( "quench", slot.quench );
+    jo.read( "brew_time", slot.brewtime );
+    jo.read( "addiction_potential", slot.addict );
+    jo.read( "fun", slot.fun );
+    jo.read( "stim", slot.stim );
+    jo.read( "healthy", slot.healthy );
+
+    if( jo.read( "spoils_in", slot.spoils ) ) {
+        slot.spoils *= 600; // JSON specifies hours so convert to turns
```
Comment (Coolthulhu) [inline src/item_factory.cpp]: `HOURS(x)` macro is better than hardcoding a multiplier. If length of an hour ever changes (for example, due to turn being shortened to 1 second, like Kevin said he wants it to work), hardcoded multipliers may be missed.

File: src/itype.cpp [hunk 3]
```diff
@@ -87,26 +87,4 @@ std::string ammo_name( std::string const &t )
...
-
-int it_comest::get_calories() const
-{
-    if( nutr >= 0 ) {
-        return ( int )( nutr * kcal_per_nutr );
-    }
-
-    return kcal;
-}
+}
```
Comment (Coolthulhu) [inline src/itype.cpp]: I'm not exactly sure why, but newlines at the end of the file are "more correct" than no endlines.

File: src/itype.h [hunk 4]
```diff
@@ -57,6 +57,52 @@ struct islot_tool {
     unsigned char turns_per_charge = 0;
 };
 
+struct islot_comestible
+{
+    /** subtype, eg. FOOD, DRINK, MED */
+    std::string comesttype;
```
Comment (Coolthulhu) [inline src/itype.h]: Could (should) become an enum in the future.

File: src/iuse.cpp [hunk 5]
```diff
@@ -1104,12 +1098,12 @@ int iuse::plantblech(player *p, item *it, bool, const tripoint &pos)
...
         }
-        const auto food = dynamic_cast<const it_comest*>(it->type);
+
         //reverses the harmful values of drinking fertilizer
-        p->mod_hunger(p->nutrition_for(food) * multiplier);
-        p->mod_thirst(-food->quench * multiplier);
-        p->mod_healthy_mod(food->healthy * multiplier, food->healthy * multiplier);
-        p->add_morale(MORALE_FOOD_GOOD, -10 * multiplier, 60, 60, 30, false, food);
+        p->mod_hunger( p->nutrition_for( it->type ) * multiplier );
+        p->mod_thirst( -it->type->comestible->quench * multiplier);
```
Comment (Coolthulhu) [inline src/iuse.cpp]: In the `caff` above, you check for comestible type. Here you don't. Fertilizer is more likely to become non-comestible than coffee.

File: src/npc.cpp [hunk 6]
```diff
@@ -1811,13 +1809,11 @@ int npc::value( const item &it, int market_price ) const
...
     // ALSO TODO: Bionics hunting from relevant factions
-    if( fac_has_job(FACJOB_DRUGS) && it.is_food() &&
-        (dynamic_cast<const it_comest*>(it.type))->addict >= 5 ) {
+    if( fac_has_job(FACJOB_DRUGS) && it.is_food() && it.type->comestible->addict >= 5 ) {
         ret += 10;
     }
 
-    if( fac_has_job(FACJOB_DOCTORS) && it.is_food() &&
-        dynamic_cast<const it_comest*>( it.type )->comesttype == "MED" ) {
+    if( fac_has_job(FACJOB_DOCTORS) && it.is_food() && it.type->comestible->comesttype == "MED" ) {
```
Comment (Coolthulhu) [inline src/npc.cpp]: While it is good for compatibility and ease of porting, in the future meds probably shouldn't be checked for with `is_food`, but `is_med`.

**Pull request merged**